### PR TITLE
Constexpr Everything Part 1: Constants

### PR DESCRIPTION
**Pull request by DesWurstes**
I do this by hand, so make sure to report those whom I didn't see.

### CODE REVIEW

**Comment by achow101** [comment]
Could you add some rationale as to why this is necessary? Code style changes need to significantly improve the developer experience and avoid serious programming bugs. I do not see how this change does either of those things, so NACK

**Comment by JeremyRubin** [comment]
Concept Ack -- making these constexpr allows us to add more static_asserts (and similar things) about relationships between constants.

File: src/zmq/zmqpublishnotifier.cpp [hunk 1]
```diff
@@ -12,10 +12,10 @@
 
 static std::multimap<std::string, CZMQAbstractPublishNotifier*> mapPublishNotifiers;
 
-static const char *MSG_HASHBLOCK = "hashblock";
-static const char *MSG_HASHTX    = "hashtx";
-static const char *MSG_RAWBLOCK  = "rawblock";
-static const char *MSG_RAWTX     = "rawtx";
+static constexpr char *MSG_HASHBLOCK = "hashblock";
```
Comment (practicalswift) [inline src/zmq/zmqpublishnotifier.cpp]: Starting with C++11 this conversion is no longer allowed. Please cast.

Applies also to the three lines below.

File: src/validation.h [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2010 Satoshi Nakamotoconstexpr
```
Comment (practicalswift) [inline src/validation.h]: I'm sure Satoshi loves `constexpr` – but this is taking it too far :-)

File: src/zmq/zmqpublishnotifier.cpp (hunk 1 above)
Comment (practicalswift) [reply src/zmq/zmqpublishnotifier.cpp]: @DesWurstes I think you might have misunderstood my comment.

My point was that simply replacing `const` with `constexpr` for `MSG_HASHBLOCK` does not give the result you want to achieve in this case :-)

More specifically you don't get a read-only variable.

What you have in `master`:

```
[cling]$ static const char * MSG_HASHBLOCK = "hashblock";
[cling]$ MSG_HASHBLOCK[0] = 'H';
error: read-only variable is not assignable
```

With your suggested change you get:

```
[cling]$ static constexpr char * MSG_HASHBLOCK = "hashblock";
warning: ISO C++11 does not allow conversion from string literal to 'char *const' [-Wwritable-strings]
[cling]$ MSG_HASHBLOCK[0] = 'H';
Segmentation fault
```

What I think you want:

```
[cling]$ static constexpr const char *MSG_HASHBLOCK = "hashblock";
[cling]$ MSG_HASHBLOCK[0] = 'H';
error: read-only variable is not assignable
```

**Comment by practicalswift** [comment]
@JeremyRubin We can use `static_assert(…)`:s on the integer constants also in the absence of `constexpr` :-)

```
[cling]$ const int i = 0;
[cling]$ static_assert(i == 0, "");
[cling]$ int j = 0;
[cling]$ static_assert(j == 0, "");
error: static_assert expression is not an integral constant expression
```

No need for `constexpr` in for that :-)

**Comment by JeremyRubin** [comment]
@practicalswift I think that is compiler dependent?

In any case, you also want to have relations be defined (e.g., constexpr functions).

**Comment by fanquake** [comment]
@DesWurstes can you fix up the whitespace issues (and squash your fixup commits) so the tests will run

**Comment by DesWurstes** [comment]
I understand that it is controversial to constexpr the constants as it's not a part of the code style. Should I open a separate Pull Request that replaces constexprs with consts?

**Comment by practicalswift** [comment]
@DesWurstes AFAIK there are no problems with `constexpr` from a code style perspective: I would say that we prefer `contexpr`:s for constants in new code. Have you gotten any contradicting information? :-)

Regarding this PR I think any objections would be based on the code churn vs benefit trade-off.

**Comment by laanwj** [comment]
>I would say that we prefer contexpr:s for constants in new code

There's no mention of that. Before doing this, the developer notes need to be updated *when* to use `constexpr` and when `const`, along with the rationale. Not everyone is sufficiently up to date with C++11 to know this—is it *always* better to use constexpr? and if so, why does `const` still exist?

The word `constexpr` doesn't appear in `doc/developer-notes.md`.

**Comment by MarcoFalke** [comment]
There hasn't been a follow-up to address the feedback (Document the new guideline in the developer notes), so I am closing this for now. Please let me know when you start working on this again, so I can reopen.

**Pull request closed without merging**